start to showing deprecated warnings

saeideng opened the pull request:
TODO:

- [x]  enable showing deprecated warning

- [x] finding disabled  test classes (not show/throw deprecated warning in testCases)

saeideng committed:
enable showing deprecated warning

saeideng commented:
error_reporting()  should return 32767 by default in test case?

markstory commented:
We had originally planned on leaving deprecated errors off as having them on will require tests be updated sooner which is better, but also makes including tests for the deprecated features harder. Do you have an idea of how we could go about ensuring that deprecated features have tests?

saeideng commented:
Goal of this PR is different 
I want to ensure all tests using new method
This is make easy to migrate to 4.x

But for ensure of old code still works
Maybe by adding `@`

 `$this->assertFalse(@controller->name('article')); `
Codes after this should be works IMO
Let me to test it

saeideng commented:
If `@` works we no longer need to new test case method for old style
This make time and easy to implement

markstory commented:
I know that @ will work but it is a bit messy, as it doesn't force us to make the tests for old code easier to delete later. If we create separate test methods now. Clean-up later will be very easy.

saeideng commented:
we need add
```
    public static function setUpBeforeClass()
    {
        parent::setUpBeforeClass();
        self::$errorLevel = error_reporting();
    }
    public static function tearDownAfterClass()
    {
        parent::tearDownAfterClass();
        error_reporting(self::$errorLevel);
    }
```
into `src/TestSuite/TestCase.php`
to ensure that `error_reporting` is `E_ALL` in all test case
but how we can applied these to just core/tests ?
I want no change/no apply into  user/tests  codes 
how can we  add this ?

markstory commented:
>  but how we can applied these to just core/tests ? I want no change/no apply into user/tests codes how can we add this ?

We don't have a way to do this right now. I think your `deprecated()` method is a cleaner approach though. It allows us to use error_reporting to `E_ALL` and test deprecated features inside the helper method.

saeideng commented:
No
I need E_ALL for current tests  not for replaced tests
Ill think a bit to find  a way

saeideng commented:
Maybe by checking class_exist  for `core/tests/....`

saeideng commented:
User apps not have core  tests by default ,true?

markstory commented:
>  I need E_ALL for current tests not for replaced tests

Couldn't that be defined in CakePHP's `phpunit.xml`?

saeideng commented:
> ouldn't that be defined in CakePHP's phpunit.xml

for local yes
my `phpunit.xml`
```php
    <php>
        <ini name="memory_limit" value="-1"/>
        <ini name="apc.enable_cli" value="1"/>
        <!-- E_ALL & ~E_USER_DEPRECATED (16383)-->
        <!-- E_ALL (32767) -->
        <ini name="error_reporting" value="32767"/>
    </php>
```
I put `debug(error_reporting());` into some of test classes 
see result 
```php
./tests/TestCase/Auth/BasicAuthenticateTest.php (line 66)
########## DEBUG ##########
(int) 32767
###########################
/tests/TestCase/BasicsTest.php (line 37)
(int) 32767
/tests/TestCase/Cache/CacheTest.php (line 76)
(int) 32767
tests/TestCase/Collection/CollectionTest.php (line 56)
(int) 32767
tests/TestCase/Console/TaskRegistryTest.php (line 74)
(int) 32767
tests/TestCase/Controller/ControllerTest.php (line 298)
(int) 16383
tests/TestCase/Core/FunctionsTest.php (line 28)
(int) 16383
tests/TestCase/Datasource/RulesCheckerTest.php (line 32)
(int) 16383
tests/TestCase/Error/ExceptionRendererTest.php (line 208)
(int) -1
tests/TestCase/Event/EventTest.php (line 38)
(int) -1
tests/TestCase/Filesystem/FolderTest.php (line 72)
(int) -1
tests/TestCase/Form/SchemaTest.php (line 49)
(int) -1
tests/TestCase/Http/ServerTest.php (line 64)
(int) -1
tests/TestCase/I18n/TimeTest.php (line 91)
(int) -1
tests/TestCase/I18n/TimeTest.php (line 91)
(int) -1
tests/TestCase/Log/LogTraitTest.php (line 37)
(int) -1
tests/TestCase/Mailer/MailerTest.php (line 30)
(int) -1
tests/TestCase/Network/SocketTest.php (line 55)
(int) -1
tests/TestCase/Routing/RouterTest.php (line 66)
(int) -1
tests/TestCase/Shell/ServerShellTest.php (line 44)
(int) -1
tests/TestCase/TestSuite/TestSuiteTest.php (line 32)
(int) -1
tests/TestCase/Utility/XmlTest.php (line 73)
(int) -1
tests/TestCase/Validation/ValidatorTest.php (line 33)
(int) -1
tests/TestCase/Database/ConnectionTest.php (line 86)
(int) -1
tests/TestCase/ORM/TableTest.php (line 137)
(int) -1
tests/TestCase/Database/ConnectionTest.php (line 86)
(int) -1
tests/TestCase/ORM/TableTest.php (line 137)
(int) -1

There were 2 errors:
1) Cake\Test\TestCase\Cache\CacheTest::testRegistry
Use Cache::getRegistry() and Cache::setRegistry() instead. - [internal], line: ??

2) Cake\Test\TestCase\Cache\CacheTest::testRegistrySet
Use Cache::getRegistry() and Cache::setRegistry() instead. - [internal], line: ??
```

after adding `setUpBeforeClass` into `src/TestSuite/TestCase.php` all above debug return `(int) 32767`

saeideng commented:
I found a way
one of invalid usage of error_reporting that cause this problem 
https://github.com/cakephp/cakephp/blob/151a7e4d6a437602c734b63b91d5c49ee106b4a0/tests/TestCase/Controller/Component/RequestHandlerComponentTest.php#L1108-L1115

Ill push a few commits to solving main problem and for this testcase as soon

saeideng commented:
with  latest commit value of `error_reporting` is `32767` for all case now

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -88,6 +94,25 @@ public function skipIf($shouldSkip, $message = '')
+        $class = get_called_class();
+        if (substr($class, 0, 18) !== $needle) {
+            return;
+        }
```
Why do you kneed to know the precise class?

saeideng replied:
For not  apply to user code/class
By this if we can ensure this setting just use in core tests
Probably for no break of user tests

saeideng replied:
I means no apply to user test cases

markstory commented on src/TestSuite/TestCase.php:
```diff
@@ -88,6 +94,25 @@ public function skipIf($shouldSkip, $message = '')
+        if (substr($class, 0, 18) !== $needle) {
+            return;
+        }
+        if ($restore) {
```
Instead of a modal parameter it would be nice to have 2 methods as it will read better.

saeideng replied:
In that case we need 3 method
1)check for tests is for core not user
2)getter
3)setter

markstory replied:
Could we update the teats that change error reporting so that they restore the old level?

saeideng replied:
We no check changes of error reporting within testCases 
So if some of tests make change this level and if no restore the level to default (within that test) this cause change of error level in during of running tests(see output of my debug(error_reporting) in above
And after that some of test's error ignored 
We need save error level before the tests and restore it after running every test
We want to make an isolation test

saeideng replied:
We must save/restore it even with using deprecated() helper 
We no a way to restore error level if we expect a exception for example `testAddInputTypeException`
resore line not execute anymore 
And we should  manage it with `setUp` `teardown`

markstory replied:
Yes. The tests that modify error_reporting() should be updated to restore the previous level. We have to do this for other process level state like the current locale.

saeideng committed:
use deprecated helper

saeideng committed:
remove unexecutable line

remove unexecutable/unneeded line
when using @expectedException

saeideng committed:
remove error_reporting() handling/use deprecated()

remove error_reporting() from

setUpBeforeClass()
tearDownAfterClass()
tearDown()

+
used deprecated() helper

saeideng commented:
finished
is there  other problem?

saeideng commented on src/TestSuite/TestCase.php:
```diff
@@ -135,6 +161,29 @@ public function tearDown()
+    public static function tearDownAfterClass()
+    {
+        parent::tearDownAfterClass();
+        self::_updateErrorLevel(true);
```
Yes
For when we expect an exception (except deprecated) within  `deprecated()`

saeideng replied:
Like `testAddInputTypeException` in above ,when you want to use `deprecated()`

saeideng replied:
I thnk a bit
Probably we abe to remove this within `teardownAfterClass()` or `tearDown()` (just on of this) 
if you concern about performance
This is not related to using/not using `deprecated()`

markstory replied:
I think we should update the tests to use try/catch or have the deprecated helper catch exceptions, restore the error level and the rethrow the exception.

saeideng replied:
We cant not try/catch for all tests at the moment 
And this way make disable all the error and is hard to update all tests
We can try/catch and rethrow within helper method (i tested this in local )
But this good for replaced tests not for current tests
So we must restore error level  on the `tearDown()` even by using good helper method for ensuring all tests(especially current tests ) no change the error level
What is problem by restore error level?
Performace?
If yes ,i can remove this method in `setUp()`
And if nedded in `tearDownAfterClass`

saeideng replied:
If you want remove all `error_reporting()` from all tests
This is not possible IMO

saeideng replied:
Do you think about removing all the calling  `error_reporting`

markstory replied:
I don't think we will be able to remove all calls to error_reporting() in the short term. I want to avoid adding more setup/teardown logic in the base class as it applies to every app/plugin as well. Instead we can apply setup/teardown to the tests that meed it now. Longer term we can fix those tests or start using deprecated().

saeideng replied:
>I want to avoid adding more setup/teardown logic in the base class as it applies to every app/plugin as well.

```php
+        $needle = 'Cake\Test\TestCase';
+        $class = get_called_class();
+        if (substr($class, 0, 18) !== $needle) {
+            return;
+        }
```

markstory replied:
I'm not sure that is something we should be doing. I agree it will work, but I would rather take the time to fix the tests now.

saeideng commented:
> I would rather take the time to fix the tests now.

I found tests that want change the error level when `error_reporting` is '32767' and '16383'
```php
//error_reporting 32767
'class:  Cake\Test\TestCase\Controller\Component\RequestHandlerComponentTest  ---method: testAddInputTypeException'
'class:  Cake\Test\TestCase\Core\FunctionsTest  ---method: testDeprecationWarningLevelDisabled'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testHandleErrorDebugOn'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testErrorMapping with data set #0'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testErrorSuppressed'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testHandleErrorDebugOff'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testHandleErrorLoggingTrace'
'class:  Cake\Test\TestCase\TestSuite\TestCaseTest  ---method: testDeprecatedWithException'

//error_reporting 16383
'class:  Cake\Test\TestCase\Controller\ControllerTest  ---method: testDeprecatedViewPropertySetterMessage with data set #0'
'class:  Cake\Test\TestCase\Controller\ControllerTest  ---method: testDeprecatedViewPropertyGetterMessage with data set #0'
'class:  Cake\Test\TestCase\Core\FunctionsTest  ---method: testDeprecationWarningEnabled'
'class:  Cake\Test\TestCase\Core\FunctionsTest  ---method: testDeprecationWarningEnabledDefaultFrame'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testHandleErrorDebugOn'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testErrorMapping with data set #0'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testErrorSuppressed'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testHandleErrorDebugOff'
'class:  Cake\Test\TestCase\Error\ErrorHandlerTest  ---method: testHandleErrorLoggingTrace'
```
if I fix these problem where possible ,this PR can be merge ?
a question, if we no want to use `Cake\TestSuite\TestCase::updateErrorLevel();` 
how we can fix https://github.com/cakephp/cakephp/blob/3.next/tests/TestCase/Core/FunctionsTest.php#L46-#L56

markstory commented:
Couldn't the `FunctionsTest` case capture/restore the error level in its setup/teardown?

saeideng commented:
like this?
https://github.com/cakephp/cakephp/blob/3.next/tests/TestCase/Controller/ControllerTest.php#L263-L279

markstory commented:
@saeideng Yes.

saeideng commented:
you say we should add these in **every** testClasses ? and remove parent codes?
this is a bit risky for future codes/tests . no?
when we forget enable/adds these codes

saeideng commented:
this is very hard to remember 
for example 
https://github.com/cakephp/cakephp/blob/151a7e4d6a437602c734b63b91d5c49ee106b4a0/tests/TestCase/Controller/Component/RequestHandlerComponentTest.php#L1108-L1115
we want change this to 
```php
     * @expectedException \Cake\Core\Exception\Exception
     * @return void
     */
    public function testAddInputTypeException()
    {
        $this->deprecated(function(){
                      $this->RequestHandler->addInputType('csv', ['I am not callable']);
        })
    }
```
in this case `error_reporting($errorLevel);` line in 
https://github.com/cakephp/cakephp/blob/9caa2e0bb1169a000a016f60fb839715420841ff/src/TestSuite/TestCase.php#L96-L102
can not be works 
so tests after this test receive incorrect error level

markstory commented:
> you say we should add these in every testClasses ? and remove parent codes?

Yes.

>  when we forget enable/adds these codes

Those tests will likely not work, which will indicate that there is a problem with the tests/code.

> this is very hard to remember

Yes but the pain is short lived for us and manually managing it makes it clearer which tests are manipulating error levels.

I think the `expectedException` annotation can still be used if the deprecated method looked like

```php
function deprecated($callback)
{
  try {
   $restore = error_reporting();
   error_reporting(E_ALL ^ E_USER_DEPRECATED);
   $callback();
  } finally {
    error_reporting($restore);
  } 
}
```

saeideng commented:
@markstory
We have problem because we no have exclusive parent class
What you think?

saeideng commented:
With try{} we ignore other excetion
We should rethrow  but I'm not sure we able to cover(rethrow) all other error/exception

markstory commented:
We can capture and rethrow all exceptions by catching \Exception. I can update the deprecated method if you would like.

saeideng commented:
I'll update deprecated helper in new PR 
I wrote some code before
Let me to try 
I concern about something, I should test it

saeideng commented:
1)we can add other method `withErrorReporting()` helper like `deprecated()` (https://github.com/cakephp/cakephp/pull/11309)  to cover tests that needs to use `error_reporting()`
usage : 
```php
    /**
     * @expectedException PHPUnit\Framework\Error\Deprecated 
     * @expectedExceptionMessage This is going away - [internal], line: ?? 
     */ 
    public function testDeprecationWarningEnabled() 
    { 
        $this->withErrorReporting(E_ALL,function (){
                deprecationWarning('This is going away', 1); 
        });
    }
```
by this we can remove all usage of `error_reporting()` from core tests 

2)we can use also bellow code for every tests that needs error_reporting()
```php
        try {
            $errorLevel = error_reporting();
            error_reporting(error level );
            //code here 
        } finally {
            error_reporting($errorLevel);
        }
```

I like 1 approach

--If you agree I'll add new helper 

--if no I'll remove all  `error_reporting()` from `setUP()`, `tearDown()` ,`tearDownAfter/BeforeCLass()` 
and instead I use 2 approach for tests that needs

saeideng commented:
managing  `error_reporting()` within `setUP()`, `tearDown()` ,`tearDownAfter/BeforeCLass()`  for every class  is not easy 
but I can to it ,my concern is about forget it for new usage of `error_reporting` in future 
other option adding trait or parent class
```php
namespace Cake\Test\TestSuite;

use Cake\TestSuite\TestCase as AppTestCase;

/**
 * Cake internall TestCase class
 */
 class TestCase extends AppTestCase
{

  private static $errorLevel;
  public static function setUpBeforeClass()
  {
      parent::setUpBeforeClass();
      self::$errorLevel = error_reporting();
  }
    public function setUp()
    {
        parent::setUp();

    }
    public function tearDown()
    {
        parent::tearDown();
        error_reporting(self::$errorLevel);
    }
    public function deprecated(){
      ...
    }
}
///test class
// use Cake\TestSuite\TestCase; <-- removing this
use Cake\Test\TestSuite\TestCase; // <-- adding this
class ControllerTest extends TestCase
{
  
}
```

markstory commented:
The `withErrorReporting` helper sounds like a simple way to wrap up tests that require changing the error level that won't require setup/teardown logic in each class :+1:

saeideng committed:
wrap tests by error reporting helpers

saeideng commented:
looks finished !!
please review

markstory commented:
Great work @saeideng :clap:
